consensus/src/creation/creator.rs

```diff
@@ -2,87 +2,134 @@ use crate::{
+
+        if self.n_candidates < threshold {
+            return Err(ConstraintError::NotEnoughParents);
+        }
+        if self.candidates.get(node_ix).is_none() {
+            return Err(ConstraintError::MissingOwnParent);
+        }
+        Ok(&self.candidates)
+    }
+}
+
+fn create_unit<H: Hasher>(
+    node_ix: NodeIndex,
+    parents: NodeMap<H::Hash>,
+    round: Round,
+) -> (PreUnit<H>, Vec<H::Hash>) {
+    let control_hash = ControlHash::new(&parents);
+    let parent_hashes = parents.into_values().collect();
+
+    let new_preunit = PreUnit::new(node_ix, round, control_hash);
+    (new_preunit, parent_hashes)
+}
 
 pub struct Creator<H: Hasher> {
-    node_id: NodeIndex,
+    round_collectors: Vec<UnitsCollector<H>>,
+    node_ix: NodeIndex,
     n_members: NodeCount,
-    candidates_by_round: Vec<NodeMap<H::Hash>>,
-    n_candidates_by_round: Vec<NodeCount>, // len of this - 1 is the highest round number of all known units
 }
 
 impl<H: Hasher> Creator<H> {
-    pub fn new(node_id: NodeIndex, n_members: NodeCount) -> Self {
+    pub fn new(node_ix: NodeIndex, n_members: NodeCount) -> Self {
         Creator {
-            node_id,
+            node_ix,
             n_members,
-            candidates_by_round: vec![NodeMap::with_size(n_members)],
-            n_candidates_by_round: vec![NodeCount(0)],
+            round_collectors: vec![UnitsCollector::new(n_members)],
         }
     }
 
     pub fn current_round(&self) -> Round {
-        (self.n_candidates_by_round.len() - 1) as Round
+        (self.round_collectors.len() - 1) as Round
     }
 
-    // initializes the vectors corresponding to the given round (and all between if not there)
-    fn init_round(&mut self, round: Round) {
-        if round > self.current_round() {
-            let new_size = (round + 1).into();
-            self.candidates_by_round
-                .resize(new_size, NodeMap::with_size(self.n_members));
-            self.n_candidates_by_round.resize(new_size, NodeCount(0));
-        }
+    // gets or initializes unit collectors for a given round (and all between if not there)
+    fn get_collector_for_round(&mut self, round: Round) -> &mut UnitsCollector<H> {
+        let round = usize::from(round);
+        if round >= self.round_collectors.len() {
+            let new_size = round + 1;
+            self.round_collectors
+                .resize(new_size, UnitsCollector::new(self.n_members));
+        };
+        &mut self.round_collectors[round]
     }
 
     /// Returns `None` if a unit cannot be created.
     /// To create a new unit, we need to have at least floor(2*N/3) + 1 parents available in previous round.
     /// Additionally, our unit from previous round must be available.
-    pub fn create_unit(&self, round: Round) -> Option<(PreUnit<H>, Vec<H::Hash>)> {
-        if !self.can_create(round) {
-            return None;
+    pub fn create_unit(&self, round: Round) -> Result<(PreUnit<H>, Vec<H::Hash>)> {
```

Comment:
I don't like the signature of this function vs how it is used (repeatedly called in a loop). This suggests that we are getting errors all the time, even though these "errors" are fully expected. Also the `MissingCollector` is the only true error -- this happening would mean there is a critical bug in the implementation, which should be distinguishable from the remaining results. 
I propose the following:
- call this function `try_create_preunit`
- make its output type `Result<CreationAttempt>`
- the `CreationAttempt` would be an enum with variants 
`Ready(PreUnit<H>, Vec<H::Hash>)`
`WaitingForEnoughParents`
`WaitingForOwnParent`
- the only error this could return is `MissingCollector`
- in the place where this function is called, log the error on `warn` or `error` level instead of `debug`, but don't log  `WaitingForEnoughParents` or `WaitingForOwnParent`, because they are not errors.

Reply:
I think the current version is fine. Of course what you propose makes sense, but I would need to give it more thought in order to tell if it's worth it... Anyway, feel free to resolve this comment.